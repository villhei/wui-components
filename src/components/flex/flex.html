<style>
  .flex {
    display: flex;
  }

  .inline-flex {
    display: inline-flex;
  }

  .raised {
    padding: 1rem;
    box-shadow: 0 0 0.25rem hsla(0, 0%, 0%, var(--small-shadow-opacity));
    border-radius: 0.25rem;
  }
</style>
<div><slot></slot></div>
