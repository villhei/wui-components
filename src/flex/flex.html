<style>
  .flex {
    display: flex;
    flex-direction: var(--flex-direction);
    flex-wrap: var(--flex-wrap);
    position: var(--position);
  }

  .inline-flex {
    display: inline-flex;
    flex-direction: var(--flex-direction);
    flex-wrap: var(--flex-wrap);
  }

  .raised {
    padding: var(--content-panel-padding);
    margin: var(--content-panel-margin);

    box-shadow: 0 0 0.25rem hsla(0, 0%, 0%, var(--small-shadow-opacity));
    border-radius: var(--border-radius);
    background-color: var(--background-color);
  }
</style>
<div><slot></slot></div>
