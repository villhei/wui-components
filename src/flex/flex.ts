import templateContent from "bundle-text:./flex.html";
import { WUIBase } from "../core";

const template = document.createElement("template");

template.innerHTML = templateContent;

export class Flex extends WUIBase {
  constructor() {
    super(template);
    this.rootNode = this.root.querySelector("div")!;
    this.update();
  }

  rootNode: HTMLDivElement;
  static get observedAttributes() {
    return ["class", "raised", "inline", "direction"];
  }

  get inline() {
    return this.getAttribute("inline") === "";
  }

  get raised() {
    return this.getAttribute("raised") === "";
  }

  get classes() {
    return this.getAttribute("class")?.split(",") || [];
  }

  get direction() {
    return this.getAttribute("direction") || "row";
  }

  update() {
    this.rootNode.style.setProperty("--flex-direction", this.direction);
    if (this.raised) {
      this.rootNode.classList.add("raised");
    } else {
      this.rootNode.classList.remove("raised");
    }
    if (this.inline) {
      this.rootNode.classList.add("inline-flex");
      this.rootNode.classList.remove("flex");
    } else {
      this.rootNode.classList.remove("inline-flex");
      this.rootNode.classList.add("flex");
    }

    this.classes.forEach((className) => this.rootNode.classList.add(className));
  }
}
