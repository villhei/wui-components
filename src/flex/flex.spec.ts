import { Flex } from "./flex";
customElements.define("wui-flex", Flex);

describe("Flex", () => {
  it("should match snapshot", () => {
    const el = new Flex();
    expect(el.shadowRoot?.innerHTML).toMatchSnapshot();
  });

  it("should set flex as display by default", () => {
    const el = new Flex();
    expect(el.style.getPropertyValue("--flex-display")).toEqual("flex");
  });

  it("should set inline-flex with inline attribute", () => {
    const el = new Flex();
    el.setAttribute("inline", "");
    expect(el.style.getPropertyValue("--flex-display")).toEqual("inline-flex");
  });

  it("should set raised global class", () => {
    const el = new Flex();
    el.setAttribute("raised", "");
    expect(el.classList).toContain("__wui-raised");
  });

  it("should allow control of wrap", () => {
    const el = new Flex();
    el.setAttribute("wrap", "wrap-reverse");
    expect(el.style.getPropertyValue("--flex-wrap")).toEqual("wrap-reverse");
  });

  it("should allow control of direction", () => {
    const el = new Flex();
    el.setAttribute("direction", "column-reverse");
    expect(el.style.getPropertyValue("--flex-direction")).toEqual(
      "column-reverse"
    );
  });
});
