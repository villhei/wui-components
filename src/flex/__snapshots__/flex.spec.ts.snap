// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex should match snapshot 1`] = `
"<style>
  :host {
    display: var(--flex-display);
    flex-direction: var(--flex-direction);
    flex-wrap: var(--flex-wrap);
  }
</style>
<slot></slot>
"
`;
